credentials/handler.go

```diff
@@ -0,0 +1,267 @@
+	return &Handler{r: r}
+}
+
+func (h *Handler) RegisterAdminRoutes(admin *x.RouterAdmin) {
+	admin.PUT(RouteItemCredentials, h.updateCredentials)
+}
+
+type AdminUpdateIdentityCredentialsBody struct {
+
+	// State is the identity's state.
+	//
+	// required: true
+	ImportCredentials []Body `json:"import_credentials"`
+}
+
+type Body struct {
+	Type   identity.CredentialsType `json:"type"`
+	Config json.RawMessage          `json:"config"`
+}
+
+// swagger:route PUT /identities/{id}/credentials v0alpha1 adminUpdateCredentials
+//
+// Update Identity Credentials
+//
+// Calling this endpoint updates the credentials according to the specification provided.
+//
+// Learn how identities work in [Ory Kratos' User And Identity Model Documentation](https://www.ory.sh/docs/next/kratos/concepts/identity-user-model).
+//
+//     Produces:
+//     - application/json
+//
+//     Schemes: http, https
+//
+//     Responses:
+//       204: emptyResponse
+//       404: jsonError
+//       500: jsonError
+func (h *Handler) updateCredentials(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	var ur AdminUpdateIdentityCredentialsBody
+	if err := errors.WithStack(jsonx.NewStrictDecoder(r.Body).Decode(&ur)); err != nil {
+		h.r.Writer().WriteError(w, r, err)
+		return
+	}
+
+	id := x.ParseUUID(ps.ByName("id"))
+	i, err := h.r.PrivilegedIdentityPool().GetIdentityConfidential(r.Context(), id)
+	if err != nil {
+		h.r.Writer().WriteError(w, r, err)
+		return
+	}
+
+	for _, v := range ur.ImportCredentials {
+		creds, ok := i.GetCredentials(v.Type)
+
+		h.r.Logger().
+			WithRequest(r).
+			WithField("id", id).
+			WithField("identity", i).
+			WithField("creds", creds).
+			Info("creds before updating - REMOVE ME AFTER TESTING - SECURITY")
+
+		if !ok {
+			creds, err = h.CreateCreds(r.Context(), v)
+		} else {
+			err = h.UpdateCreds(r.Context(), id, creds, v)
+		}
+		if err != nil {
+			h.r.Writer().WriteError(w, r, err)
+		}
+		h.r.Logger().
+			WithRequest(r).
+			WithField("id", id).
+			WithField("identity", i).
+			WithField("creds", creds).
+			Info("Setting updated creds - REMOVE ME AFTER TESTING - SECURITY")
+
+		i.SetCredentials(v.Type, *creds)
+	}
+
+	if err := h.r.IdentityManager().Update(
+		r.Context(),
+		i,
+		identity.ManagerAllowWriteProtectedTraits,
+	); err != nil {
+		h.r.Writer().WriteError(w, r, err)
+		return
+	}
+
+	h.r.Writer().Write(w, r, i)
+}
+
+func uid(provider oidc.ProviderCredentialsConfig) string {
+	return fmt.Sprintf("%s:%s", provider.Provider, provider.Subject)
+}
+
+func (h *Handler) CreateCreds(ctx context.Context, newCreds Body) (*identity.Credentials, error) {
+	if newCreds.Type == identity.CredentialsTypeOIDC {
+		var newConfig oidc.CredentialsConfig
+		if err := json.Unmarshal(newCreds.Config, &newConfig); err != nil {
+			return nil, err
+		}
+		for _, provider := range newConfig.Providers {
+			i, _, err := h.r.PrivilegedIdentityPool().FindByCredentialsIdentifier(ctx, identity.CredentialsTypeOIDC, uid(provider))
```

Comment:
shouldn't this check if the identity we're trying to update has those credentials already and continue?